### DISCUSSION TITLE
Commit timestamps in GitHub don't match repo

### DISCUSSION THREAD

**Discussion by mscheper** [community/community#22695] [New to GitHub]
<p>I have various old git repos, including one or two converted from older repo managers (e.g. Perforce). They include commit comments from many years ago—some of the commits are older than a couple of the people I work with! But now that these projects are in GitHub, the commit dates are completely wrong. If I type <code>git log</code> for a file at the command line, and the most recent commit date is from several years ago, the same commit in GitHub has a much more recent timestamp.</p>
<p>The way I went about it was to create a new (private) repo in GitHub, type <code>git remote add origin git@github.com:myaccount/mynewgithubrepo.git</code>, and <code>git push -u origin master</code>. Nothing fancy.</p>
<p>Does GitHub just not like old dates, or is there data somewhere in my git repos that might not necessarily match what’s in the git log? I know I’ve used <code>git commit --date</code> for some commits, which look correct in <code>git log</code>, but not in GitHub. Are there any tricks to get the dates in GitHub to match what’s in the git log? Because the git log dates are correct, as far as I’m concerned, and they’re the ones I want to see in GitHub.</p>
<p>Thanks,<br>
MS.</p>

**Comment by nethgato** [comment]
<p>Hey hey <a class="mention" href="/mscheper">@mscheper</a>! I guess I’m curious what you <em>do</em> see in GitHub? Do you see dates for these commits as the time they were committed? Or an even less accurate, irrelevant date?</p>
<p>If it’s displaying the date it was committed, and not the historic commit date you see in <code>commit log</code>, you’ll want to make sure the author date, matches the historic value.</p>
<p>I want to elevate <a href="https://docs.github.com/en/github/setting-up-and-managing-your-github-profile/troubleshooting-commits-on-your-timeline#how-github-uses-the-git-author-date-and-commit-date">our relevant docs</a>, but also, <a href="https://stackoverflow.com/questions/454734/how-can-one-change-the-timestamp-of-an-old-commit-in-git">this ol SO post</a>.</p>
<p>Between these two, you should have what you need, but the remote add might not be updating both author and commit dates from your history.</p>
<p>If there is anything more GitHub-specific that isn’t answered in those threads, follow up and I’ll do what I can!</p>

**Comment by mscheper** [comment, marked as answer]
<p>I didn’t realise there was a distinction between the ‘author date’ and ‘commit date’. It does seem odd that GitHub would choose to show a different date from what’s in the logs, but I’m sure some thought went into that decision, and I know a lot of people are more comfortable with history being rewritten than I. :slightly_smiling_face:</p>
<p>I fixed it with these commands:</p>
<pre><code class="lang-auto">git filter-branch --env-filter 'export GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"'
git push -f
</code></pre>
<p>GitHub doesn’t seem bothered by dates older than it is, either. :wink:</p>
<p>Thanks for your help!</p>

**Reply by MarcellHarmaci** [reply]
For me this messed up all the commits before branching from the main branch. After force pushing GitHub didn't recognise the main and feature branch's common history. The message was:
> There isn't anything to compare.

Instead of this I had to rebase with keeping the original commit times, then force push.
```
git rebase --committer-date-is-author-date main feature
git push -f
```
Only after these steps could I create the pull request.

**Reply by juliamcneill** [reply]
To anyone using the initial answer here, I would also suggest that you figure how many commits you actually need to and should rewrite. I made the mistake of running this on a long commit history, when I only needed to fix the dates of the last few commits.

The command that worked for me is:
```
git filter-branch --env-filter 'export GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"' {SHA of commit before first commit}..HEAD
```

**Reply by BretHudson** [reply]
Expanding on @juliamcneill's edit, if you also want to make sure the update commits are signed, you can add a `--commit-filter` to the command as well:

```
git filter-branch --env-filter 'export GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"' --commit-filter 'git commit-tree -S "$@"' {SHA of commit before first commit}..HEAD
```

If you're running into issues with a "Cannot create a new backup" command, modify it with `-f` before the filters:

```
git filter-branch -f --env-filter 'export GIT_COMMITTER_DATE="$GIT_AUTHOR_DATE"' --commit-filter 'git commit-tree -S "$@"' {SHA of commit before first commit}..HEAD
```

**Reply by ammarshah** [reply]
In case someone is feeling a bit sceptical using `filter-branch`, here are the alternate commands using `filter-repo`.

1. If you want to rewrite `CommitDate`s of all the commits:

```shell
# Using filter-branch
git filter-branch --env-filter '
  GIT_COMMITTER_DATE=$GIT_AUTHOR_DATE
'

# Using filter-repo
git filter-repo --commit-callback '
  commit.committer_date = commit.author_date
'
```

2. If you want to rewrite `CommitDate` of a specific commit:

```shell
# Using filter-branch
git filter-branch --env-filter '
  if test "$GIT_COMMIT" = "<SHA_of_commit>"
  then
    GIT_COMMITTER_DATE=$GIT_AUTHOR_DATE
  fi
'

# Using filter-repo
git filter-repo --commit-callback '
  if commit.original_id == b"<SHA_of_commit>":
    commit.committer_date = commit.author_date
'
```

3. If you want to rewrite `CommitDate`s of a range of commits:

```shell
# Using filter-branch
git filter-branch --env-filter '
  GIT_COMMITTER_DATE=$GIT_AUTHOR_DATE
' <SHA_of_commit_made_before_the_one_you_want_to_rewrite>..master

# or three commits from the top
git filter-branch --env-filter '
  GIT_COMMITTER_DATE=$GIT_AUTHOR_DATE
' master~3..master

# Using filter-repo
git filter-repo \
  --refs <SHA_of_commit_made_before_the_one_you_want_to_rewrite>..master \
  --commit-callback '
    commit.committer_date = commit.author_date
  '
# or three commits from the top
git filter-repo \
  --refs master~3..master \
  --commit-callback '
    commit.committer_date = commit.author_date
  '
```

You can read more details [here](https://gist.github.com/ammarshah/5d4e38b3d69c24f6e6cc5b62e2ea2a5a).

**Comment by nethgato** [comment]
<p>Not a problem at all! I’m glad I could be helpful :bow:</p>
